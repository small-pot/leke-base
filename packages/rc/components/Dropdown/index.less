@import "../../less/theme.less";
@keyframes slide-open {
  0%{
    transform: scaleY(0.8);
    opacity: 0;
  }
  100%{
    transform: scaleY(1);
    opacity: 1;
  }
}
@keyframes slide-close {
  0%{
    transform: scaleY(1);
    opacity: 1;
  }
  100%{
    transform: scaleY(0.8);
    opacity: 0;
  }
}
.leke-slide-open,.leke-slide-close{
  animation-duration:350ms;
  animation-timing-function:cubic-bezier(0.23,1,0.32,1);
}
.leke-slide-open{
  animation-name:slide-open;
}
.leke-slide-close{
  animation-name:slide-close;
}
.leke-trigger{
  display: inline-block;
  outline: none;
}
.leke-dropdown{
  position: absolute;
  z-index:999;
  background-color: @background-color;
  border-radius: @radius;
}
.leke-dropdown-direction-down{
  box-shadow: @shadow-down;
  transform-origin: 0 0;
}
.leke-dropdown-direction-up{
  box-shadow: @shadow-up;
  transform-origin: 0 100%;
}
.leke-hide{
  display: none;
}