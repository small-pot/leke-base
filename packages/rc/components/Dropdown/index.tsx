import React, {ReactNode, CSSProperties, useLayoutEffect, useRef, useEffect} from 'react';
import {createPortal} from 'react-dom';
import classNames from 'classnames';
import {useAnimation,useControl} from '@leke/hooks';

function getPosition(trigger:HTMLElement,container:HTMLElement) {
    let left=0,top=0;
    let offsetElement=trigger;
    while (offsetElement!==container){
        left+=offsetElement.offsetLeft;
        top+=offsetElement.offsetTop;
        offsetElement=offsetElement.offsetParent as HTMLElement;
    }
    return {
        left,
        top,
        bottom:top+trigger.offsetHeight,
        right:left+trigger.offsetWidth
    };
}

export interface dropdownProps {
    visible?:boolean,
    onVisibleChange?:(boolean)=>void,
    getPopupContainer:(HTMLElement)=>HTMLElement,
    trigger:ReactNode,
    triggerStyle?:CSSProperties,
    triggerClassName?:string,
    popup:ReactNode,
    popupStyle?:CSSProperties,
    popupClassName?:string,
    placement:'bottomLeft'|'bottomCenter'|'bottomRight'|'topLeft'|'topCenter'|'topRight',
    triggeredEvent:('focus'|'hover')[]
}
export interface childProps {
    ref:React.RefObject<HTMLDivElement>
    tabIndex?:number,
    className:string,
    style?:CSSProperties,
    onFocus?:(e)=>void,
    onMouseDown?:(e)=>void,
    onMouseEnter?:(e)=>void,
    onMouseLeave?:(e)=>void,
    onBlur?:(e)=>void
}
export default function Dropdown(props:dropdownProps) {
    const {trigger,triggerClassName,triggerStyle,getPopupContainer,popup,popupClassName,popupStyle,placement,triggeredEvent}=props;
    const [visible,setVisible]=useControl(props.visible,props.onVisibleChange);
    const portalContainerRef=useRef<HTMLElement>(null);
    const popupRef=useRef(null);
    const triggerRef=useRef(null);
    if(visible&&!portalContainerRef.current&&typeof window==='object'){
        portalContainerRef.current=getPopupContainer(triggerRef.current);
    }
    const enterClassName=placement.indexOf('bottom')===0?'leke-slide-down':'leke-slide-up';
    const leaveClassName=placement.indexOf('bottom')===0?'leke-slide-up':'leke-slide-down';
    useAnimation({
        ref:popupRef,
        open:visible,
        enterClassName,
        leaveClassName,
        afterLeaveClassName:'leke-hide'
    });
    useLayoutEffect(()=>{
        if(!visible){
            return;
        }
        const portalContainer=portalContainerRef.current;
        if(portalContainer){
            const triggerElement=triggerRef.current;
            const container=popupRef.current;
            if(portalContainer!==document.body&&window.getComputedStyle(portalContainer).position==='static'){
                portalContainer.style.position='relative';
            }
            const {left,top,bottom,right}=getPosition(triggerElement,portalContainer);
            const containerHeight=container.offsetHeight;
            switch (placement) {
            case "bottomLeft":
                container.style.top = bottom + 'px';
                container.style.bottom='';
                container.style.left = left + 'px';
                container.style.right='';
                break;
            case "bottomRight":
                container.style.top = bottom + 'px';
                container.style.bottom='';
                container.style.left='';
                container.style.right=portalContainer.offsetWidth-right+'px';
                break;
            case "bottomCenter":
                container.style.top = bottom + 'px';
                container.style.bottom='';
                container.style.left = (triggerElement.offsetWidth-container.offsetWidth)/2+left + 'px';
                container.style.right='';
                break;
            case "topLeft":
                container.style.top = top  - containerHeight + 'px';
                container.style.bottom='';
                container.style.left = left + 'px';
                container.style.right='';
                break;
            case "topRight":
                container.style.top = top  - containerHeight + 'px';
                container.style.bottom='';
                container.style.left='';
                container.style.right=portalContainer.offsetWidth-right+'px';
                break;
            case "topCenter":
                container.style.top = top  - containerHeight + 'px';
                container.style.bottom='';
                container.style.left = (triggerElement.offsetWidth-container.offsetWidth)/2+left + 'px';
                container.style.right='';
                break;
            }
        }
    },[triggerRef,portalContainerRef,popupRef,visible,placement]);
    function show() {
        if(popupRef.current&&popupRef.current.className.indexOf(leaveClassName)!==-1){
            return;
        }
        setVisible(true);
    }
    function hide() {
        setVisible(false);
    }
    const triggerProps:childProps={
        ref:triggerRef,
        className:classNames('leke-trigger',triggerClassName),
        style:triggerStyle
    };
    const popupProps:childProps={
        ref:popupRef,
        className:classNames('leke-dropdown',popupClassName),
        style:popupStyle
    };
    if(triggeredEvent.indexOf('focus')!==-1){
        triggerProps.onFocus=show;
        triggerProps.onBlur=hide;
        triggerProps.tabIndex=-1;
        popupProps.onMouseDown=(e)=>e.preventDefault();
    }
    if(triggeredEvent.indexOf('hover')!==-1){
        triggerProps.onMouseEnter=show;
        triggerProps.onMouseLeave=hide;
    }
    return (
        <span {...triggerProps}>
            {trigger}
            {portalContainerRef.current?createPortal(
                <div style={{position:'absolute',top:0,left:0,width:'100%'}}>
                    <div {...popupProps}>{popup}</div>
                </div>,
                portalContainerRef.current
            ):null}
        </span>
    );
}
Dropdown.defaultProps={
    getPopupContainer:()=>document.body,
    placement:'bottomLeft',
    triggeredEvent:['focus']
};