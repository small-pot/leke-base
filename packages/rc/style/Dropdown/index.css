@keyframes slide-down {
  0% {
    transform: scaleY(0.8);
    opacity: 0;
  }
  100% {
    transform: scaleY(1);
    opacity: 1;
  }
}
@keyframes slide-up {
  0% {
    transform: scaleY(1);
    opacity: 1;
  }
  100% {
    transform: scaleY(0.8);
    opacity: 0;
  }
}
.leke-slide-down,
.leke-slide-up {
  animation-duration: 350ms;
  animation-timing-function: cubic-bezier(0.23, 1, 0.32, 1);
}
.leke-slide-down {
  transform-origin: 0 0;
  animation-name: slide-down;
}
.leke-slide-up {
  transform-origin: 100% 0;
  animation-name: slide-up;
}
.leke-trigger {
  display: inline-block;
  outline: none;
}
.leke-dropdown {
  position: absolute;
  z-index: 999;
  background-color: #ffffff;
  border-radius: 4px;
}
.leke-dropdown-direction-down {
  box-shadow: 0px 5px 12px 4px rgba(0, 0, 0, 0.08), 0px 3px 6px 0px rgba(0, 0, 0, 0.12), 0px 1px 2px -2px rgba(0, 0, 0, 0.12);
}
.leke-dropdown-direction-up {
  box-shadow: 0px -5px 12px 4px rgba(0, 0, 0, 0.08), 0px -3px 6px 0px rgba(0, 0, 0, 0.12), 0px -1px 2px -2px rgba(0, 0, 0, 0.12);
}
.leke-hide {
  display: none;
}
