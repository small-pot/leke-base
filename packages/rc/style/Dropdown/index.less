@import "../../less/animation.less";
@import "../../less/theme.less";
@keyframes slide-down {
  0%{
    transform: scaleY(0.8);
    opacity: 0;
  }
  100%{
    transform: scaleY(1);
    opacity: 1;
  }
}
@keyframes slide-up {
  0%{
    transform: scaleY(1);
    opacity: 1;
  }
  100%{
    transform: scaleY(0.8);
    opacity: 0;
  }
}
.leke-slide-down,.leke-slide-up{
  animation-duration:350ms;
  animation-timing-function:cubic-bezier(0.23,1,0.32,1);
}
.leke-slide-down{
  transform-origin: 0 0;
  animation-name:slide-down;
}
.leke-slide-up{
  transform-origin: 100% 0;
  animation-name:slide-up;
}
.leke-trigger{
  display: inline-block;
  outline: none;
}
.leke-dropdown{
  position: absolute;
  z-index:999;
  background-color: @background-color;
  border-radius: @radius;
}
.leke-dropdown-direction-down{
  box-shadow: @shadow-down;
}
.leke-dropdown-direction-up{
  box-shadow: @shadow-up;
}
.leke-hide{
  display: none;
}