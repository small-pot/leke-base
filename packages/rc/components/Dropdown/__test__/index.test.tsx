import React from "react";
import Dropdown from "../";
import '@testing-library/jest-dom/extend-expect';
import {render,screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const defaultProps={
    trigger:<div style={{width:200}}>trigger</div>,
    popup:<div>popup</div>
};
describe('Dropdown ', function() {
    it('test default',async function () {
        const {container} = render(
            <Dropdown
                {...defaultProps}
            />
        );
        const trigger=container.querySelector('.leke-trigger');
        userEvent.click(trigger);
        expect(trigger).toHaveFocus();
        expect(screen.getByText('popup')).toBeVisible();
        userEvent.click(document.body);
        expect(screen.getByText('popup').parentElement).toHaveClass('leke-slide-close');
    });
    it('test hover',async function () {
        const {container} = render(
            <Dropdown
                {...defaultProps}
                triggeredEvent={['hover']}
            />
        );
        const trigger=container.querySelector('.leke-trigger');
        userEvent.hover(trigger);
        expect(screen.getByText('popup')).toBeVisible();
        userEvent.unhover(trigger);
        expect(screen.getByText('popup').parentElement).toHaveClass('leke-slide-close');
    });
    it('test placement',async function () {
        const {container,rerender} = render(
            <Dropdown
                {...defaultProps}
                placement='bottomLeft'
            />
        );
        const trigger=container.querySelector('.leke-trigger');
        userEvent.click(trigger);
        const popupContainer=screen.getByText('popup').parentElement;
        expect(popupContainer).toHaveClass('leke-dropdown-direction-down');
        const popup=popupContainer.parentElement;
        expect(popup).toMatchSnapshot();
        rerender(
            <Dropdown
                {...defaultProps}
                placement='topLeft'
            />
        );
        expect(popupContainer).toHaveClass('leke-dropdown-direction-up');
        expect(popup).toMatchSnapshot();
        rerender(
            <Dropdown
                {...defaultProps}
                placement='bottomCenter'
            />
        );
        expect(popup).toMatchSnapshot();
        rerender(
            <Dropdown
                {...defaultProps}
                placement='bottomRight'
            />
        );
        expect(popup).toMatchSnapshot();
        rerender(
            <Dropdown
                {...defaultProps}
                placement='topCenter'
            />
        );
        expect(popup).toMatchSnapshot();
        rerender(
            <Dropdown
                {...defaultProps}
                placement='topRight'
            />
        );
        expect(popup).toMatchSnapshot();
    });
});