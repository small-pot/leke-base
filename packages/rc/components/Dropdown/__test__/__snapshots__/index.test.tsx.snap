// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown  test placement 1`] = `
<div
  style="position: absolute; top: 0px; left: 0px; width: 100%;"
>
  <div
    class="leke-dropdown leke-dropdown-direction-down leke-slide-open"
    style="min-width: 0; top: 0px; left: 0px;"
  >
    <div>
      popup
    </div>
  </div>
</div>
`;

exports[`Dropdown  test placement 2`] = `
<div
  style="position: absolute; top: 0px; left: 0px; width: 100%;"
>
  <div
    class="leke-dropdown leke-dropdown-direction-up"
    style="min-width: 0; top: 0px; left: 0px;"
  >
    <div>
      popup
    </div>
  </div>
</div>
`;

exports[`Dropdown  test placement 3`] = `
<div
  style="position: absolute; top: 0px; left: 0px; width: 100%;"
>
  <div
    class="leke-dropdown leke-dropdown-direction-down"
    style="min-width: 0; top: 0px; left: 0px;"
  >
    <div>
      popup
    </div>
  </div>
</div>
`;

exports[`Dropdown  test placement 4`] = `
<div
  style="position: absolute; top: 0px; left: 0px; width: 100%;"
>
  <div
    class="leke-dropdown leke-dropdown-direction-down"
    style="min-width: 0; top: 0px; left: 0px;"
  >
    <div>
      popup
    </div>
  </div>
</div>
`;

exports[`Dropdown  test placement 5`] = `
<div
  style="position: absolute; top: 0px; left: 0px; width: 100%;"
>
  <div
    class="leke-dropdown leke-dropdown-direction-up"
    style="min-width: 0; top: 0px; left: 0px;"
  >
    <div>
      popup
    </div>
  </div>
</div>
`;

exports[`Dropdown  test placement 6`] = `
<div
  style="position: absolute; top: 0px; left: 0px; width: 100%;"
>
  <div
    class="leke-dropdown leke-dropdown-direction-up"
    style="min-width: 0; top: 0px; left: 0px;"
  >
    <div>
      popup
    </div>
  </div>
</div>
`;
